static/app/utils/profiling/profile/chromeTraceProfile.tsx

```diff
@@ -0,0 +1,290 @@
+  endQueue: ChromeTrace.Event[]
+): 'B' | 'E' {
+  if (!beginQueue.length && !endQueue.length) {
+    throw new Error('Profile contains no events');
+  }
+
+  const nextBegin = beginQueue[0];
+  const nextEnd = endQueue[0];
+
+  if (!nextEnd) {
+    return 'B';
+  }
+  if (!nextBegin) {
+    return 'E';
+  }
+  if (nextBegin.ts < nextEnd.ts) {
+    return 'B';
+  }
+  if (nextEnd.ts < nextBegin.ts) {
+    return 'E';
+  }
+  return 'B';
+}
+
+function buildProfile(
+  processId: string,
+  threadId: string,
+  events: ChromeTrace.Event[]
+): EventedProfile {
+  let processName: string = `pid (${processId})`;
+  let threadName: string = `tid (${threadId})`;
+
+  // We dont care about other events besides begin, end, instant events and metadata events
+  const timelineEvents = events.filter(
+    e => e.ph === 'B' || e.ph === 'E' || e.ph === 'X' || e.ph === 'M'
+  );
+
+  // @TODO use a heap so we dont need to sort this again afterwards
+  const beginQueue: Array<ChromeTrace.Event> = [];
+  const endQueue: Array<ChromeTrace.Event> = [];
+
+  for (let i = 0; i < timelineEvents.length; i++) {
+    const event = timelineEvents[i];
+
+    // M events are not pushed to the queue, we just store their information
+    if (event.ph === 'M') {
+      if (event.name === 'thread_name' && typeof event.args.name === 'string') {
+        threadName = `${event.args.name} (${threadId})`;
+        continue;
+      }
+
+      if (event.name === 'process_name' && typeof event.args.name === 'string') {
+        processName = `${event.args.name} (${processId})`;
+        continue;
+      }
+    }
+
+    // B, E and X events are pushed to the timeline. We transform all X events into
+    // B and E event, so that they can be pushed onto the queue and handled
+    if (event.ph === 'B') {
+      beginQueue.push(event);
+      continue;
+    }
+
+    if (event.ph === 'E') {
+      endQueue.push(event);
+      continue;
+    }
+
+    if (event.ph === 'X') {
+      if (typeof event.dur === 'number' || typeof event.tdur === 'number') {
+        beginQueue.push({...event, ph: 'B'});
+        endQueue.push({...event, ph: 'E', ts: event.ts + (event.dur ?? event.tdur ?? 0)});
+        continue;
+      }
+    }
+  }
+
+  beginQueue.sort(chronologicalSort);
+  endQueue.sort(chronologicalSort);
+
+  if (!beginQueue.length) {
+    throw new Error('Profile does not contain any frame events');
+  }
+
+  const firstTimestamp = beginQueue[0].ts;
+
+  if (typeof firstTimestamp !== 'number') {
+    throw new Error('First begin event contains no timestamp');
+  }
+
+  const profile = new EventedProfile(
+    0,
+    0,
+    0,
+    `${processName}: ${threadName}`,
+    'milliseconds'
+  );
+
+  const stack: ChromeTrace.Event[] = [];
+  const frameCache = new Map<string, Frame>();
+
+  while (beginQueue.length > 0 || endQueue.length > 0) {
+    const next = getNextQueue(beginQueue, endQueue);
+
+    if (next === 'B') {
+      const item = beginQueue.shift();
```

Comment:
I figure this might make a difference with larger profiles but because of this `shift()`, this loops is now `O(n^2)`. A simple fix can be to either just advance an index without shifting the array or simply sort the array in reverse chronological order and use `pop()` instead.

Reply:
No, it's a solid call. I'll reverse it!